language/move-lang/src/errors/diagnostic_codes.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use codespan_reporting_new::diagnostic::Severity;
+
+//**************************************************************************************************
+// Main types
+//**************************************************************************************************
+
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub(crate) struct DiagnosticInfo {
+    pub(crate) severity: Severity,
+    category: Category,
+    code: u8,
+    message: &'static str,
+}
+
+pub(crate) trait DiagnosticCode: Copy {
+    fn severity() -> Severity;
+
+    fn category() -> Category;
+
+    fn code_and_message(self) -> (u8, &'static str);
+
+    fn into_info(self) -> DiagnosticInfo {
+        let severity = Self::severity();
+        let category = Self::category();
+        let (code, message) = self.code_and_message();
+        DiagnosticInfo {
+            severity,
+            category,
+            code,
+            message,
+        }
+    }
+}
+
+//**************************************************************************************************
+// Categories and Codes
+//**************************************************************************************************
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
+#[repr(u8)]
+pub enum Category {
+    // Uncategorized = 0,
+    // Syntax = 1,
+    NameResolution = 2,
+    // TypeSafety = 3,
+    // AbilitySafety = 4,
+    // ReferenceSafety = 5,
+    // MoveSafety = 6,
+}
+
+// pub enum SyntaxError {
+// }
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
+#[repr(u8)]
+pub enum NameResolutionError {
+    UnboundModule = 1,
+}
+
+// pub enum TypeSafetyError {}
+
+// pub enum AbilitySafetyError {}
+
+// pub enum ReferenceSafetyError {}
+
+// pub enum MoveSafetyError {}
+
+//**************************************************************************************************
+// impls
+//**************************************************************************************************
+
+impl DiagnosticInfo {
+    pub fn render(self) -> (/* code */ String, /* message */ &'static str) {
+        let Self {
+            severity,
+            category,
+            code,
+            message,
+        } = self;
+        let sev_prefix = match severity {
+            Severity::Error => "E",
+            Severity::Warning => "W",
+            _ => unimplemented!(),
+        };
+        let cat_prefix: u8 = category as u8;
+        debug_assert!(cat_prefix <= 99);
+        let string_code = format!("{}{:02}{:03}", sev_prefix, cat_prefix, code);
+        (string_code, message)
+    }
+}
+
+impl DiagnosticCode for NameResolutionError {
+    fn severity() -> Severity {
+        Severity::Error
+    }
+
+    fn category() -> Category {
+        Category::NameResolution
+    }
```

Comment:
Might be cleaner to have these as associated constants in the trait definition.

Reply:
Oh is that a thing? Neet